### PR TITLE
cql3/type_json: throw when converted double is out of range

### PR DESCRIPTION
**Pull request by tchaikov**
cql3/type_json: throw when converted double is out of range

before this change, a 64-bit float point number is compared with
an int64_t integer directly, but C++ compiler implicitly promoted
the integer to a nearest float to compare them. because the design
of IEEE 754, a 64-bit float point number cannot always represent
the value of an int64_t integer. when performing conversion,
`std::numeric_limits<int64_t>::max()`, i.e., 9223372036854775807,
is converted the nearest float point number of 9223372036854775808.F.
and we use it as the upper bound of range of valid values convertable
to int64_t. if the number being converted is actually
9223372036854775808.F. we'd consider it in the range. but it is not.

in this change, we compare these two 64-bit numbers in a more
sophiscated way by checking the numbers being compared with the
`std::numeric_limits<T>::max()`.

this change addresses a corner case which only happens when
serializing a double numbers sitting at the boundary to a JSON
node of int64 type.

Signed-off-by: Kefu Chai <kefu.chai@scylladb.com>

### CODE REVIEW

File: cql3/type_json.cc [hunk 1]
```diff
@@ -119,7 +119,8 @@ template <typename T> static T to_int(const rjson::value& value) {
                 "for int64 type: {} (it should not contain fractional part {})", value, fractional));
         }
 
-        if (std::numeric_limits<T>::min() > double_value || double_value > std::numeric_limits<T>::max()) {
+        if (static_cast<double>(std::numeric_limits<T>::min()) > double_value ||
```
Comment (avikivity) [inline cql3/type_json.cc:122]: The warning may mean the change is wrong. What if double_value was indeed 9223372036854775808? Then the compare will fail, and we'll allow the cast, even though it doesn't fit in int64_t.

Please fix and add a unit test for int64_t::max{-1,+0,+1} and similar for the minimum value.

Comment (avikivity) [reply cql3/type_json.cc:122]: s/change/original code/. The change doesn't change correctness, just silences the warning.

**Comment by tchaikov** [comment]
sure. will fix it instead of covering the issue.

**Comment by tchaikov** [comment]
changelog:

- compare the 64-bit numbers in the right way
- a test case is added.


@avikivity i updated the change. could you please take another look?

File: cql3/type_json.cc [hunk 2]
```diff
@@ -75,6 +77,31 @@ static sstring quote_json_string(const sstring& value) {
     return oss.str();
 }
 
+template <std::floating_point Tf, std::integral Ti>
+static std::partial_ordering cmp(Tf x, Ti y) {
+    return x <=> y;
+}
+
+// we are not able to fit an uint64_t or a 64 bit float into a wider numeric type
```
Comment (nyh) [inline cql3/type_json.cc:85]: It took me a while to understand the purpose of this template because the comment starts with the implementation detail instead of the goal for the function. I would start the comment with something like: 
"The following function cmp()  compares a floating-point and an integer number. Because there is no "wider" type to which a 64-bit integer and 64-bit floating point can be casted to do this comparison, we need to take a more convoluted approach:"

File: cql3/type_json.cc [hunk 3]
```diff
@@ -75,6 +77,31 @@ static sstring quote_json_string(const sstring& value) {
...
+// to compare them, so need to do this in a more convoluted way. specifically, we
+// have to compare the float with the maximum value represented by the integer
+// type to see if the nearest conversion changes the result of the comparison.
+//
+// see https://github.com/JuliaLang/julia/blob/49b361aa2ec68efe4cb0d5870e772eb3e0d89993/base/float.jl#L553-L565
+template <std::floating_point Tf, std::integral Ti>
+requires (sizeof(Tf) <= 8 && sizeof(Ti) >= 8)
```
Comment (nyh) [inline cql3/type_json.cc:92]: What in the code below requires size 8? Isn't just sizeof(Tf) <= sizeof(Ti) enough?

File: cql3/type_json.cc [hunk 4]
```diff
@@ -75,6 +77,31 @@ static sstring quote_json_string(const sstring& value) {
...
+static std::partial_ordering cmp(Tf x, Ti y) {
+    auto fy = static_cast<Tf>(y);
+    if (x < fy || (x == fy && (fy < static_cast<Tf>(std::numeric_limits<Ti>::max()) &&
+                               std::bit_cast<Ti>(fy) < y))) {
```
Comment (nyh) [inline cql3/type_json.cc:96]: I don't understand the role of the various pieces of this logic (and in particular what this bit_cast does). A comment explaining or giving examples would have helped, but perhaps not necessary if you have good tests covering these exact code paths. I'll write more below about the tests.

**Review by nyh — CHANGES_REQUESTED** [review]
Please create an issue explaining what the user-visible bugs caused by this problem is - this is important to decide if this fix needs to be backported. However, before creating a new issue, please check if this is the same as one I already created - https://github.com/scylladb/scylladb/issues/10137 and if not, explain in the new issue how it differs. Please also check out the test for that issue -  https://github.com/scylladb/scylladb/commit/b650ff5808e3b742be2894378d26aed1f6ca5b33 - does your patch fix that test, or not?

The next thing I want to be improved in this PR is the testing:

1.  Because I couldn't wrap my had about the correctness of the cmp() function, I would have liked to see a unit test specifically for this function (this will probably require it not be static, and having a longer name) and a test that covers the various cases that the different `||`, `&&` in the function cover.
2. I would have liked to see a CQL test (in test/cql-pytest or test/cql) which checks the real user-visible problem (the same examples I want you to also explain in the issue). E.g., did you see a problem in 9223372036854775808.0? Please also test the same thing on Cassandra and note if Cassandra has the same bug or it's unique to us. By the way, I added a test for #10137 in https://github.com/scylladb/scylladb/commit/b650ff5808e3b742be2894378d26aed1f6ca5b33 - is this also a test for this issue? Does the patch fix it?
3. Finally, I wasn't thrilled by the test that you did add. First, it's a unit test for some function, to_json_object, and as a read I have no idea what is this function and who uses it - and I would like to see an explanation. But more importantly, the test has tons of code which is even more complex than the code it is testing, to the point you need a test for the test :-( Can't you have a much simpler test with constants? Please also write in comments or the commit message if/how the test failed before the patch.

Also, when you create an issue (or discover that #10137 is the same?) please refer to it in the test comments, and also add a Fixes to it in the commit message.

Thanks!

**Comment by nyh** [comment]
By the way, unfortunately we have several other bugs in our JSON conversion functions in CQL, some of which are in floating point. You can see the various tests marked "xfail" in `test/cql-pytest/test_json.py` (tests that I wrote, heavily commented) and in `test/cql-pytest/cassandra_tests/validation/entities/json_test.py (tests I translated from Cassandra's unit tests). 
If you or someone else wants to look into fixing all those bugs (some of which are probably related), I would be very happy.

File: cql3/type_json.cc (hunk 3 above)
Comment (tchaikov) [reply cql3/type_json.cc:92]: because 64bit is the problem i am trying to fix. if the LHS is a 32-bit float point number, and the RHS is a 32-bit integer, we can just cast them to a 64 bit float point numbers and compare them.

File: cql3/type_json.cc (hunk 2 above)
Comment (tchaikov) [reply cql3/type_json.cc:85]: good point. my head was full of the IEEE754 things when working on this, and really wanted to dump them in the form of code or document. will revise the comment as you suggest.

**Comment by tchaikov** [comment]
> Please create an issue explaining what the user-visible bugs caused by this problem is - this is important to decide if this fix needs to be backported. However, before creating a new issue, please check if this is the same as one I already created - #10137 and if not, explain in the new issue how it differs. Please also check out the test for that issue - [b650ff5](https://github.com/scylladb/scylladb/commit/b650ff5808e3b742be2894378d26aed1f6ca5b33) - does your patch fix that test, or not?
> 
> The next thing I want to be improved in this PR is the testing:
> 
> 1. Because I couldn't wrap my had about the correctness of the cmp() function, I would have liked to see a unit test specifically for this function (this will probably require it not be static, and having a longer name) and a test that covers the various cases that the different `||`, `&&` in the function cover.

i will try to explain this in plain English. to generate the different test cases for all the cases would take me even more time and would result in even more length test code, i think.

> 2. I would have liked to see a CQL test (in test/cql-pytest or test/cql) which checks the real user-visible problem (the same examples I want you to also explain in the issue). E.g., did you see a problem in 9223372036854775808.0? Please also test the same thing on Cassandra and note if Cassandra has the same bug or it's unique to us. By the way, I added a test for #10137 in [b650ff5](https://github.com/scylladb/scylladb/commit/b650ff5808e3b742be2894378d26aed1f6ca5b33) - is this also a test for this issue? Does the patch fix it?

i will try. but that'd take much longer. is it a must to include this fix in scylla? or we can have the test later.

> 3. Finally, I wasn't thrilled by the test that you did add. First, it's a unit test for some function, to_json_object, and as a read I have no idea what is this function and who uses it - and I would like to see an explanation. But more importantly, the test has tons of code which is even more complex than the code it is testing, to the point you need a test for the test :-( Can't you have a much simpler test with constants? Please also write in comments or the commit message if/how the test failed before the patch.
> 

i really tried. but i think, if i cannot make it simpler, i think i should at least hold the standard of maintainability and readability. the "tons of code" is designed to generate the dataset. simpler tests also require

- less magic numbers
- more readable code

constants hurts the readability. one can hardly find the relation between 9223372036854775807, 9223372036854775808 and the width of this number, not to mention deducing the intention of the test from these numbers. guess that's why human beings have an intention to name everything. more importantly, unlike fixed point number, one cannot just use `a_float_number + 1` to get the "next" great float number.

> Also, when you create an issue (or discover that #10137 is the same?) please refer to it in the test comments, and also add a Fixes to it in the commit message.

will do. thought it was a corner case though..

> 
> Thanks!

File: cql3/type_json.cc (hunk 3 above)
Comment (nyh) [reply cql3/type_json.cc:92]: Does this code actually work for things which aren't specifically double and int_64_t?
E.g., if x is 32-bit (float) and y is 128-bit integer, does this code actually work? I'm not sure I understand why.  It seems like it checks "x < fy" but the 128-bit integer doesn't fit the floating point so... I don't understand what the code does. Again, a unit test might have convinced me it's correct.

Comment (tchaikov) [reply cql3/type_json.cc:92]: float32 versus. int128 should work. if this is something we support, will surely add tests for it. was just reluctant  to test an internal helper.. even if we will make a public one for testing it.

**Comment by nyh** [comment]
> > 1. Because I couldn't wrap my had about the correctness of the cmp() function, I would have liked to see a unit test specifically for this function (this will probably require it not be static, and having a longer name) and a test that covers the various cases that the different `||`, `&&` in the function cover.
> 
> i will try to explain this in plain English. to generate the different test cases for all the cases would take me even more time and would result in even more lengthy test code, i think.

I don't understand how you can trust the code without such a unit test. Avi suggested in his review that maybe the result for 9223372036854775808 is wrong. How can you know if it's wrong or not without actually testing this number and seeing what cmp() does with in practice?

Regarding the length of the test code - the problem is not the just length, but the difficulty of understanding. For example, even you had tests that do lines like

```
CHECK(cmp(1234.5, 1234, std::partial_ordering::greater))
```

Then this test will be trivial to understand even if it has a 100 lines like this.
The problem I had with your test is that it had 100 lines of real floating-point logic, logic which itself needs to be tested or at least understood - which I didn't.
>

> > 2. I would have liked to see a CQL test (in test/cql-pytest or test/cql) which checks the real user-visible problem (the same examples I want you to also explain in the issue). E.g., did you see a problem in 9223372036854775808.0? Please also test the same thing on Cassandra and note if Cassandra has the same bug or it's unique to us. By the way, I added a test for [INSERT JSON of bigint (64-bit) in scientific notation is not accurate above 2^53 #10137](https://github.com/scylladb/scylladb/issues/10137) in [b650ff5](https://github.com/scylladb/scylladb/commit/b650ff5808e3b742be2894378d26aed1f6ca5b33) - is this also a test for this issue? Does the patch fix it?
> 
> i will try. but that'd take much longer. is it a must to include this fix in scylla? or we can have the test later.

Let's begin with creating an issue and explaining in it what the problem is. I still don't understand what the problem is :-( Is there a real user-visible error in CQL, e.g., somebody uses the number 9223372036854775808.0 in some sort of JSON-oriented CQL function (?) and gets a wrong result? Or what? If you understand the user-visible problem, it's trivial to write a test - I volunteer to do it for you, if you just open an issue and explain what the user-visible problem is.

> 
> > 3. Finally, I wasn't thrilled by the test that you did add. First, it's a unit test for some function, to_json_object, and as a read I have no idea what is this function and who uses it - and I would like to see an explanation.
> 
> sorry, i don't quite follow you here. an explanation for the function being tested? or the reason why we want to test it? or its callers?

It seems your code modifies a function called to_int(), and then the test checks a function calls from_json_object(). As a reviewer who don't remember all the details, I need some context - where are these functions used in practice and why are you testing the bug to_int() through to_json_object() and not directly? What does from_json_object() promise but failed to deliver before your patch? I would have also like to see some explanation on what was the bug before this patch (if you had an issue on what the problem is, the test code can say in a comment something like `// This reproduces issue #123456`.

> 
> > But more importantly, the test has tons of code which is even more complex than the code it is testing, to the point you need a test for the test :-( Can't you have a much simpler test with constants? Please also write in comments or the commit message if/how the test failed before the patch.
> 
> i really tried. but i think, if i cannot make it simpler, i think i should at least hold the standard of maintainability and readability. the "tons of code" is designed to generate the dataset. simpler tests also require
> 
>     * less magic numbers
> 
>     * more readable code
> 
> 
> constants hurts the readability. one can hardly find the relation between 9223372036854775807, 9223372036854775808 and the width of this number,

You can used named constants like `std::numeric_limits<int64_t>::max` and then `std::numeric_limits<int64_t>::max - 1`,
and name intermediate constants. Your test went an extra mile by inventing some sort if floating-point simulator, which I felt I wasn't sure was correct - and didn't know how the correctness of this simulator effects the power of the test.

> not to mention deducing the intention of the test from these numbers. guess that's why human beings have an intention to name everything. more importantly, unlike fixed point number, one cannot just use `a_float_number + 1` to get the "next" great float number.

Let's go back to basics - because I still don't understand what the problem that this PR fixes actually is - I don't understand how getting the "next great float number" helps to test it.

> > Also, when you create an issue (or discover that #10137 is the same?) please refer to it in the test comments, and also add a Fixes to it in the commit message.
> 
> will do. thought it was a corner case though..

I still don't understand which corner case it is.

**Comment by tchaikov** [comment]
@nyh Nadav, thank you for your review and comments. i think i understand your intention better now. the test and the fix comes from Avi's comment of 

> Please fix and add a unit test for int64_t::max{-1,+0,+1} and similar for the minimum value.

but you are looking at the change from user's perspective. my theory is, "if we fix a public/internal function, the outcome should be better". even without understanding the whole code path from end to end.

please note that we don't have a off-the-shelf tool for generating the float point version of "int64_t::max{-1,+0,+1}" for us. hence the test.  anyway, i will try to find out the "problem". but i need more time so i can read though the code.

**Comment by tchaikov** [comment]
@avikivity @nyh i 

- created an issue.
- commented on the newly added `cmp()` function
- added a `cql-pytest` test in `test_json.py`

but after a second thought, i don't want to expose `cmp()` as a public function just for testing it. if we apply this approach to more and more functions just because some of us find it difficult to understand a certain internal function, it would make the general code structure and interfaces more difficult to manage. i think if a certain function (in the sense of feature) can be tested with a relatively good coverage, we don't need to expose its internals for even better testability.

**Comment by tchaikov** [comment]
#10137 is not the same as #13077. as you put in the description of the former, the ideal fix should happen in RapidJSON. but with #13077, we can at least address one occurrence of  
```
runtime error: 9.22337e+18 is outside the range of representable values of type 'long'
```
please see the issue description for more details.

**Comment by nyh** [comment]
> #10137 is not the same as #13077. as you put in the description of the former, the ideal fix should happen in RapidJSON. but with #13077, we can at least address one occurrence of
> 
> ```
> runtime error: 9.22337e+18 is outside the range of representable values of type 'long'
> ```
> 
> please see the issue description for more details.

I'll take a look at your PR more carefully again, but so far I'm almost convinced that this is exactly the same issue as https://github.com/scylladb/scylladb/issues/10137 just that that issue was about the "INSERT JSON" command and your issue is about the "fromJson()" function. These are slightly different implementations but I think they have exactly the same problem which I described in #10137 - any number that is an integer but "pretends" to be a float (ends in ".0" or has "e") is read by RapidJSON into a "double", and loses precision if this number is more then `2^53`. In particular the numbers 9223372036854775807 (`2^63-1`) and 9223372036854775808 (`2^63`) are indistinguishable when assigned into a double... No matter how "clever" your comparison code is, you either refuse both of them - or accept both of them. BOTH of these decisions are wrong, and I think what your patch does is to switch from one wrong decision to a new wrong decision...

This is why I wanted to see a "real" test, e.g., trying `fromJson()` on the above two numbers, and wanted you to demonstrate that previously wrong behavior (e.g., both the above numbers are accepted) is changed into the correct behavior (only the smaller one accepted), not a different wrong behavior. The sophisticated tests which you added did not convince me of anything because I didn't understand what they check :-(

Anyway, I'll try to give you more specific comments after I review your patch again. Maybe my above comments are wrong and I misunderstood something?

File: cql3/type_json.cc [hunk 5]
```diff
@@ -75,6 +77,82 @@ static sstring quote_json_string(const sstring& value) {
...
+// - 1 sign bit
+// - 11 bits of exponent
+// - 52 bits of significand
+//
+// At the two ends of the range of the numbers represented by a 64 bit float
+// point number, the exponent is greater or equal to 1, so the float point numbers
+// are not able to represent all integers any more. Specifically, two nearest
+// representable integers greater than 2^53 would be have a spacing greater or
+// equal to 2. And, we cannot assume the existence of float128, we are not able
+// to fit an uint64_t or a 64 bit float into a wider numeric type for comparing
```
Comment (nyh) [inline cql3/type_json.cc:95]: This exposition doesn't explain to me what it means to "compare" the some floating point number to 9223372036854775807 and then compare it to 9223372036854775808. The point is, because the floating point number lost some precision you can't make a decision that it's equal one but not the other - it must be either equal both or not equal both. So I don't understand what you mean by comparing accurately or not accurately :-(
In other words, the floating point number you are trying to compare is already not accurate, there is no way to make comparison to it accurate.
What am I missing?

File: cql3/type_json.cc [hunk 6]
```diff
@@ -75,6 +77,82 @@ static sstring quote_json_string(const sstring& value) {
...
+// them, we need to develop an algorithem to tackle the problem of comparing a
+// float point number and a fixed point number, where the fixed point number is
+// so large that the former is not able to represent it accurately, because the
+// number of bits of float point number's significand is not enough for represent
+// the all the bits of a two's complement number.
+//
+// here is a hypothetical example, assuming we need to compare 4.0 with 3, and a
+// 64-bit float point number is not able to represent 3, the nearest float point
+// number to "3" is "2". the hypothetical distribution of IEEE-754 float point
```
Comment (nyh) [inline cql3/type_json.cc:104]: Can you please use a real example? I don't understand your hypothetical example, or even its notation (you start by saying that 4.0 is floating point and 3 is an int, and then you say that 2 (not 2.0) is floating point, I don't understand>).

File: test/boost/type_json_test.cc [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+  // from_json_object() should not throw at seeing: std::numeric_limits<int64_t>::max() - 1
+  //
+  // well sort of, as the 64 bit float point is not to present this number, we pick the
+  // last greatest 64 bit float point number less than std::numeric_limits<int64_t>::max() .
+  rjson::value less_than_int64_t_max{closest_to_max.previous()};
+  BOOST_CHECK_NO_THROW(from_json_object(*int64_type, less_than_int64_t_max));
+  // we don't test std::numeric_limits<int64_t>::max() here. because 64 bit IEEE-754 float
+  // point is not able to represent std::numeric_limits<int64_t>::max(),
+  // i.e., 9223372036854775807. if we cast it to double, we'll have a closest float number,
+  // i.e., 9223372036854775808.
```
Comment (nyh) [inline test/boost/type_json_test.cc:154]: Doesn't this mean that the same bug you tried to fix still exists? (this is exactly the example from https://github.com/scylladb/scylladb/issues/10137).

File: test/cql-pytest/test_json.py [hunk 8]
```diff
@@ -136,6 +136,31 @@ def test_fromjson_bigint_overflow_prepared(cql, table1):
...
+    MAX_BIGINT = 2 ** 63 - 1
+    p = unique_key_int()
+    stmt = cql.prepare(f"INSERT INTO {table1} (p, bigv) VALUES (?, fromJson(?))")
+    # all the tested number in this test case can be represented by IEEE-754
+    # 64-bit float, in order to verify that we only accept the numbers in the
+    # range of bigint without worrying if RapidJSON nearest rounds them first.
+    with pytest.raises(FunctionFailure):
+        # the last IEEE-754 64-bit float point number less than min value
+        # of 64-bit signed long
+        cql.execute(stmt, [p, f'{MIN_BIGINT - 2048}.0'])
```
Comment (nyh) [inline test/cql-pytest/test_json.py:152]: Did this test **fail** before your patch? If so, maybe I'm finally starting to understand what you tried to fix in this patch! I kept thinking it was about numbers very close to the cutoff, but this test is about numbers more far away from the cutoff and should more obviously result in an error.

If the answer is yes, please add a comment before this test explaining that it reproduces issue #NNNN, otherwise I can only guess how this specific tests relates to the issue or your fix.

Comment (nyh) [reply test/cql-pytest/test_json.py:152]: Seeing that you have here a bunch of checks one after another, it will also be helpful to have a comment "reproduces issue #NNNN" above the specific checks that failed before the patch. I assume it's not all of them?

**Comment by nyh** [comment]
https://github.com/scylladb/scylladb/pull/16625 was just committed, and supposedly solves the same issue as this PR does, so this PR should probably be closed.
**However**, this PR has a functional test, in test_json.py, that the other PR is missing, so maybe we should have a new PR, by you or @michoecho, that includes just the test_json.py test from this PR.

**Comment by tchaikov** [comment]
closing as suggested by Nadav.
